### PR TITLE
iDeep: F.split_axis

### PR DESCRIPTION
**Pull request by niboshi**
iDeep support in F.split_axis (Separated from #3865. Based on #4276.)

### CODE REVIEW

**Comment by niboshi** [comment]
This code causes FPE:
```
import chainer
from chainer import functions as F
import numpy as np

a = np.ones((10, 4, 3, 2), np.float32)

#y1, y2 = F.split_axis(a, indices_or_sections=2, axis=0)
y1, y2 = F.split_axis(a, indices_or_sections=[0, 5], axis=0)
```

```
$ CHAINER_USE_IDEEP=always python ~/bbb.py
Floating point exception (core dumped)
```
(with gdb)
```
...
Thread 9 "python" received signal SIGFPE, Arithmetic exception.
[Switching to Thread 0x7fffd3dad780 (LWP 6441)]
0x00007fffddf23d93 in mkldnn::impl::cpu::simple_reorder_impl<(mkldnn_data_type_t)1, (mkldnn_memory_format_t)1, (mkldnn_data_type_t)1, (mkldnn_memory_format_t)1, true, mkldnn::impl::cpu::spec::direct_copy_except_dim_0>::execute(mkldnn::i$pl::cpu::cpu_reorder_pd_t const*, float const*, float*) [clone ._omp_fn.212] () from /home/niboshi/anaconda/anaconda3/lib/python3.6/site-packages/ideep4py/lib/libmkldnn.so.0

(gdb) bt

#0  0x00007fffddf23d93 in mkldnn::impl::cpu::simple_reorder_impl<(mkldnn_data_type_t)1, (mkldnn_memory_format_t)1, (mkldnn_data_type_t)1, (mkldnn_memory_format_t)1, true, mkldnn::impl::cpu::spec::direct_copy_except_dim_0>::execute(mkldnn::impl::cpu::cpu_reorder_pd_t const*, float const*, float*) [clone ._omp_fn.212] () from /home/niboshi/anaconda/anaconda3/lib/python3.6/site-packages/ideep4py/lib/libmkldnn.so.0
#1  0x00007fffd44f4ac3 in __kmp_invoke_microtask () from /home/niboshi/anaconda/anaconda3/lib/python3.6/site-packages/ideep4py/lib/libiomp5.so
#2  0x00007fffd44c3257 in __kmp_invoke_task_func () from /home/niboshi/anaconda/anaconda3/lib/python3.6/site-packages/ideep4py/lib/libiomp5.so
#3  0x00007fffd44c28d5 in __kmp_launch_thread () from /home/niboshi/anaconda/anaconda3/lib/python3.6/site-packages/ideep4py/lib/libiomp5.so
#4  0x00007fffd44f4fa4 in _INTERNAL_26_______src_z_Linux_util_cpp_16f8393c::__kmp_launch_worker(void*) () from /home/niboshi/anaconda/anaconda3/lib/python3.6/site-packages/ideep4py/lib/libiomp5.so
#5  0x00007ffff7bbd7fc in start_thread (arg=0x7fffd3dad780) at pthread_create.c:465
#6  0x00007ffff78eab5f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

**Comment by niboshi** [comment]
This will be fixed in the next version of ideep soon.

**Comment by niboshi** [comment]
Rebased.

**Comment by okuta** [comment]
jenkins, test this please.

File: chainer/functions/array/split_axis.py [hunk 1]
```diff
@@ -51,6 +59,42 @@ def forward(self, inputs):
...
+    def _ideep_is_indices_or_sections_supported(self, indices_or_sections):
+        # Returns True if the value of ``indices_or_sections`` is supported
+        # by iDeep.
+        ios = indices_or_sections
+        if isinstance(ios, collections.Iterable):
+            if len(ios) == 0:
+                return False  # Empty sequence
+            if ios[0] == 0:
+                return False  # Sequence starting with 0
+            if any([ios[i-1] == ios[i] for i in six.moves.range(1, len(ios))]):
```
Comment (okuta) [inline chainer/functions/array/split_axis.py]: How about using `for` loop?

**Comment by okuta** [comment]
jenkins, test this please.

File: chainer/functions/array/split_axis.py (hunk 1 above)
Comment (niboshi) [reply chainer/functions/array/split_axis.py]: Fixed.

**Comment by okuta** [comment]
Test was failed by segmentation fault.

**Comment by niboshi** [comment]
It seems it occurs in some specific configurations:
```
from chainer import functions as F
import numpy as np


def func(conf):
    shape = conf['shape']
    indices_or_sections = conf['indices_or_sections']
    axis = conf['axis']
    a = np.ones(shape, np.float32)
    ys = F.split_axis(a, indices_or_sections, axis)
    print([_.shape for _ in ys])


oks = [
    dict(
        shape = (1, 2, 1, 1),
        indices_or_sections = 2,
        axis = 1),
    dict(
        shape = (1, 4, 1, 1),
        indices_or_sections = 2,
        axis = 1),
    dict(
        shape = (1, 16, 1, 1),
        indices_or_sections = 2,
        axis = 1),
    dict(
        shape = (8, 1, 1, 1),
        indices_or_sections = 2,
        axis = 0),
    dict(
        shape = (1, 1, 8, 1),
        indices_or_sections = 2,
        axis = 2),
    dict(
        shape = (1, 1, 1, 8),
        indices_or_sections = 2,
        axis = 3),
    dict(
        shape = (1, 8, 1, 1),
        indices_or_sections = 1,
        axis = 1),
    dict(
        shape = (1, 6, 1, 1),
        indices_or_sections = 2,
        axis = 1),
]

ngs = [
    dict(
        shape = (1, 8, 1, 1),
        indices_or_sections = 2,
        axis = 1),
    dict(
        shape = (1, 8, 1, 1),
        indices_or_sections = 4,
        axis = 1),
    dict(
        shape = (1, 8, 1, 1),
        indices_or_sections = 8,
        axis = 1),
]

# OK
for conf in oks:
    func(conf)

print("OK!")

# NG!
func(ngs[0])
#func(ngs[1])
#func(ngs[2])
```

**Comment by niboshi** [comment]
Added workaround for segfaults.

**Comment by okuta** [comment]
Did you report to iDeep project?

**Comment by niboshi** [comment]
Yes, I did.

**Comment by okuta** [comment]
OK.

**Pull request merged**